mdsuite/file_io/lammps_trajectory_files.py

```diff
@@ -57,222 +62,247 @@
+
+            # get number of configs from file length
+            file.seek(0)
+            num_lines = sum(1 for _ in file)
+            n_configs_float = num_lines / (self._n_particles + self.n_header_lines)
+            n_configs = int(round(n_configs_float))
+            assert abs(n_configs_float - n_configs) < 1e-10
+
+            # get information on which particles with which id belong to which species
+            # by analysing the first configuration
+            file.seek(0)
+            self._species_dict = self._get_species_information(
+                file, header_property_names
+            )
+
+            # extract sample step information from consecutive headers
+            file.seek(0)
+            sample_rate = self._get_sample_rate(file)
+
+        # all species have the same properties
+        properties_list = list()
+        for key, val in self._property_dict.items():
+            properties_list.append(PropertyInfo(name=key, n_dims=len(val)))
+
+        species_list = list()
+        for key, val in self._species_dict.items():
+            species_list.append(
+                SpeciesInfo(
+                    name=key,
+                    n_particles=len(val["line_idxs"]),
+                    properties=properties_list,
+                )
+            )
 
-        header = self._read_header(self.f_object)  # get the first header tensor_values
-        self.f_object.seek(0)  # go back to the start of the file
-
-        return int(header[3][0])
+        self._mdata = TrajectoryMetadata(
+            n_configurations=n_configs,
+            species_list=species_list,
+            box_l=box_l,
+            sample_rate=sample_rate,
+        )
 
-    def _get_number_of_configurations(self, number_of_atoms: int):
-        """
-        Get the number of configurations
+        self._batch_size = mdsuite.utils.meta_functions.optimize_batch_size(
+            filepath=self.file_path, number_of_configurations=n_configs
+        )
 
-        Parameters
-        ----------
-        number_of_atoms : int
-                Number of atoms in each of the trajectories
-        Returns
-        -------
+        return self._mdata
 
-        """
-        number_of_lines = line_counter(
-            self.file_path
-        )  # get the number of lines in the file
+    def get_configurations_generator(
+        self,
+    ) -> typing.Iterator[mdsuite.file_io.file_read.TrajectoryChunkData]:
+        n_configs = self._mdata.n_configurations
+        n_batches, n_configs_remainder = divmod(int(n_configs), int(self._batch_size))
 
-        return int(number_of_lines / (number_of_atoms + self.header_lines))
+        with open(self.file_path, "r") as file:
+            file.seek(0)
+            for _ in tqdm.tqdm(range(n_batches)):
+                yield self._read_process_n_configurations(file, self._batch_size)
+            if n_configs_remainder > 0:
+                yield self._read_process_n_configurations(file, n_configs_remainder)
```

Comment:
parts of this are also either identical duplicates or only differ by 

```py
# skip the header
mdsuite.file_io.file_read.skip_n_lines(file, self.n_header_lines)
```

and could also be combined in the common parent class.

Reply:
For a text file yes, for binary it cannot be so generic but I guess binary files will be exclusively handled by chemfiles and not read so manually by us.

Reply:
I will try to streamline the tabular-text-style readers.
For binary readers, it might be beneficial to not batch by time step but maybe by species or property